### ISSUE TITLE
FEATURE: Add Redux for State Management

### ISSUE THREAD

**Issue by aorinevo** [yargs/yargs#1150]
# Introduction

In certain places, yargs does state management.  There are several different ways this is done, some of which lead to tight coupling between modules.  

By adding Redux, we can consolidate and standardize the different state management implementations, resulting in cleaner; and more predictable and decoupled code.

# Migration Plan

- Create `aorinevo/I-1150_master_AN_redux` is the main feature branch
- Create a `aorinevo/I-1150_master_AN_redux_{{state key}}` branch for each piece of state that redux will handle.  Here is a tentative list of the different pieces of state and status:
    - middleware (merged into `aorinevo/I-1150_master_AN_redux`)
    - context (partial changes in `aorinevo/I-1150_master_AN_redux_context` branch)
    - usage (partial changes in `aorinevo/I-1150_master_AN_redux_usage` branch)
    - command (none)
- When `aorinevo/I-1150_master_AN_{{state key}}` is dev complete, merge into `aorinevo/I-1150_master_AN_redux`.  
- When all state branches are merged into `I-1150_master_AN_redux`, merge to `yargs/master`

 module | status | notes 
---|---|---
middleware | done | merged into `aorinevo/I-1150_master_AN_redux`
usage | done | changes in `aorinevo/I-1150_master_AN_redux_usage`
command | in-dev | No longer need to pass usage to command.js; might be able to deprecate getUsageInstance from yargs API; moved defaultCommand to redux; aliasMap is all that's left
validation | none |
yargs | none | 

# Notes

- We were able to prove out that redux can be used in a node app for state management. 
 Although this is an interesting use case, it is still not clear that it is the right tool for the job (see https://github.com/yargs/yargs/pull/1182).  With that being said, we are putting this on hold and shifting our attention to componentizing the validation module as was done for yargs-parser.
- Slim down unit tests for reducers since most of the functionality is already covered by yargs' unit tests.
- Redux v4.0.0 is 24KB unminified and 6KB minified
- Related to issue https://github.com/yargs/yargs/issues/1117 and corresponding PR https://github.com/yargs/yargs/issues/1118
- End-of-life for Node 4 was 4/30/18.
- Use undoable in place of freeze/unfreeze: https://redux.js.org/recipes/implementing-undo-history

# Questions

- Can we drop support for Node 4 in the next major yargs release?

**Comment by aorinevo** [comment]
Opened PR #1151

**Comment by aorinevo** [comment]
+ @bcoe and @evocateur

**Comment by aorinevo** [comment]
<img width="675" alt="screen shot 2018-06-13 at 8 38 27 pm" src="https://user-images.githubusercontent.com/3240137/41385459-c5af858a-6f49-11e8-96ae-e48819889e88.png">

**Comment by aorinevo** [comment]
@nexdrew, @bcoe mentioned that you worked on some of yargs’ state management code.  I am working on migrating that to redux (see PR #1151).

If you have some time, I’d like to learn more from you.

Any ideas or suggestions would be greatly appreciated.  

Thank you in advance!

**Comment by aorinevo** [comment]
I am currently working on migrating state management in `lib/usage.js` to redux in three phases.
  - Phase 1 - Complete
    - add `/actions/usage.js`
    - add `/reducers/usage.js`
    - add `usageReducer` to `lib/store.js`
    - migrate one property of state from current state management to redux (i.e. showHelpOnFail)
    - add coverage for `actions/usage.js` and `/reducers/usage.js`
  - Phase 2 - Complete
    - migrate remaining properties that are not passed by reference from current state management to redux (i.e. `failureOutput`, `epilog`, and `usageDisabled`)
  - Phase 3 - Ready for Dev
    - migrate remaining properties that **are** passed by reference from current state management to redux (i.e. `usages`, `commands`, `descriptions`, and `examples`)

**Comment by evocateur** [comment]
Awesome!

I definitely agree we should drop Node 4 in the next major.

**Comment by maniator** [comment]
Hey all

Coming in here from a conversation with @aorinevo on slack.

I am not sure what advantage using redux (which is a frontend state management library) is needed for a simple one and done command line tool?

Can the data be stored in individual instances?

**Comment by maniator** [comment]
I honestly think adding complicated redux logic into something that can be just a stored map is a bad idea

**Comment by maniator** [comment]
You can consider using an Observable in which you can store and pipe data from it whenever.

**Comment by bcoe** [comment]
@maniator @aorinevo I've definitely felt a little gun shy about pulling in all of redux, even though I do think there's some benefits to rethinking how we manage state. My concerns with yargs being:

* coupling: usage, yargs, validation, etc., all take each other as arguments.
* ability to rollback to prior state; commands are currently implemented as a top-down traversal and when a failure happens we fallback to the prior parser state in a few instances -- this logic is currently ugly and bespoke.

long story short @maniator, there are some pretty big problems around state management, and the flow of yargs is a bit more complicated than you would think -- curious what your thoughts are, what library you'd advocate instead of redux?

**Comment by evocateur** [comment]
@maniator Yargs is not that simple, sorry.

Observable... When was that going to be added to the ECMA spec, again?

Redux is "just functions," we're not pulling in anything (relatively) giant like `react-redux`.

**Comment by john-osullivan** [comment]
I came across this issue because I'm building an interactive CLI via `yargs`+`ink` which draws on some significant data sources, and Redux makes that state management easier within React.  `ink` is getting more & more popular, so I imagine that yargs and React will keep running into each other.

I don't know enough about `yargs` internals to contribute on whether it should use React within itself, but if you did, there's room for some really interesting integrations.  The first thing I'll be figuring out is making the Redux store persist to the installation location, so that data can be cached across calls.

Maybe the move isn't to build Redux directly into `yargs`, but I think there's definitely room for a library built on top of `yargs` which leverages Redux to provide a lot of easy functionality for free.  My approach is that the CLI is just another client, so I can imagine a lot of features which would support large, complex CLIs.  That `commandDir` function really is a godsend!

**Comment by aorinevo** [comment]
Closing this for now.  If there is interest in picking this up again, I'd be happy to continue the work.
